File: RTC.Console/Exercises/Chapter12.cs
using RTC.Datastructures;
using RTC.Objects;
using RTC.Primitives;
using RTC.Primitives.Patterns;

namespace RTC.Console.Exercises;

public class Chapter12 : IChapterExercise
{
    public void Run()
    {
        var world = new World();

        var floor = new Plane
        {
            Material = new Material(Colour.White)
            {
                Transparency = 0.9d,
                Reflective = 0.9d
            }
        };
        world.Shapes.Add(floor);

        var leftWall = new Plane
        {
            Transformation = Matrix.TranslationMatrix(0, 0, 5) *
                             Matrix.RotationYMatrix(-Math.PI / 4) *
                             Matrix.RotationXMatrix(Math.PI / 2) *
                             Matrix.ScalingMatrix(10, 0.01d, 10),
            Material = new Material(new CheckerPattern(Colour.White, Colour.Black,
                Matrix.ScalingMatrix(0.06d, 0.06d, 0.06d)))
        };
        world.Shapes.Add(leftWall);

        var rightWall = new Plane
        {
            Transformation = Matrix.TranslationMatrix(0, 0, 5) *
                             Matrix.RotationYMatrix(Math.PI / 4) *
                             Matrix.RotationXMatrix(Math.PI / 2) *
                             Matrix.ScalingMatrix(10, 0.01d, 10),
            Material = new Material(new CheckerPattern(Colour.White, Colour.Black,
                Matrix.ScalingMatrix(0.06d, 0.06d, 0.06d)))
        };
        world.Shapes.Add(rightWall);

        var middle = new Cube(Matrix.TranslationMatrix(0, 1, 1));
        world.Shapes.Add(middle);
        
        var left = new Cube(Matrix.TranslationMatrix(-2, 0.4d, 0.5d) * Matrix.ScalingMatrix(0.4d, 0.4d, 0.4d));
        left.Material = left.Material with { Colour = new Colour(0.8d, 0.8d, 0.2d) };
        world.Shapes.Add(left);
        
        var right = new Cube(Matrix.TranslationMatrix(2, 0.7d, 0) * Matrix.ScalingMatrix(0.7d, 0.7d, 0.7d));
        right.Material = right.Material with { Colour = new Colour(0.2d, 0.8d, 0.8d) };
        world.Shapes.Add(right);

        world.Lights.Add(new PointLight(Colour.White, Vec4.Point(-10, 10, -10)));

        var camera = new Camera(300, 300, Math.PI / 3)
        {
            Transformation = Matrix.ViewTransform(Vec4.Point(0, 2.5d, -4), Vec4.Point(0, 1, 0), Vec4.Vector(0, 1, 0))
        };

        var image = camera.Render(world);
        image.SavePng("Chapter12.png");
    }
}

File: exercises/chapter12.cpp
#include "RayTracerChallenge/datastructures/matrix.hpp"
#include "RayTracerChallenge/datastructures/vec4.hpp"
#include "RayTracerChallenge/objects/camera.hpp"
#include "RayTracerChallenge/objects/canvas.hpp"
#include "RayTracerChallenge/objects/colour.hpp"
#include "RayTracerChallenge/objects/cube.hpp"
#include "RayTracerChallenge/objects/light.hpp"
#include "RayTracerChallenge/objects/material.hpp"
#include "RayTracerChallenge/objects/patterns.hpp"
#include "RayTracerChallenge/objects/plane.hpp"
#include "RayTracerChallenge/objects/world.hpp"
#include <fstream>
#include <memory>
#include <numbers>

int main() {
  rtc::World world{};

  rtc::Plane plane{};
  auto planeMaterial = plane.getMaterial();
  planeMaterial.colour = rtc::WHITE;
  planeMaterial.transparency = 0.9f;
  planeMaterial.reflective = 0.9f;
  plane.setMaterial(planeMaterial);
  world.objects.push_back(std::make_unique<rtc::Plane>(plane));

  const rtc::Plane leftWall{
      rtc::translationMatrix(0, 0, 5) *
          rtc::rotationMatrixY(-std::numbers::pi_v<float> / 4.0f) *
          rtc::rotationMatrixX(std::numbers::pi_v<float> / 2.0f) *
          rtc::scalingMatrix(10, 0.01f, 10),
      rtc::Material{std::make_unique<rtc::CheckerPattern>(
          rtc::WHITE, rtc::BLACK, rtc::scalingMatrix(0.06f, 0.06f, 0.06f))}};
  world.objects.push_back(std::make_unique<rtc::Plane>(leftWall));

  const rtc::Plane rightWall{
      rtc::translationMatrix(0, 0, 5) *
          rtc::rotationMatrixY(std::numbers::pi_v<float> / 4.0f) *
          rtc::rotationMatrixX(std::numbers::pi_v<float> / 2.0f) *
          rtc::scalingMatrix(10, 0.01f, 10),
      rtc::Material{std::make_unique<rtc::CheckerPattern>(
          rtc::WHITE, rtc::BLACK, rtc::scalingMatrix(0.06f, 0.06f, 0.06f))}};
  world.objects.push_back(std::make_unique<rtc::Plane>(rightWall));

  world.objects.push_back(std::make_unique<rtc::Cube>(
      rtc::translationMatrix(0, 1, 1), rtc::Material{rtc::Colour{0, 1, 0}}));

  world.light = rtc::Light{rtc::WHITE, rtc::point(-10, 10, -10)};

  const rtc::Camera camera{rtc::viewTransform(rtc::point(0, 2.5f, -4.0f),
                                              rtc::point(0, 1, 0),
                                              rtc::vector(0, 1, 0)),
                           300, 300, std::numbers::pi_v<float> / 3.0f};

  const rtc::Canvas canvas = camera.render(world);
  std::ofstream file;
  file.open("image.ppm");
  file << canvas.toPPM();
  file.close();

  return 0;
}
